File: Sort/InsertionSort.cs
/*
 * Problem: InsertionSort
 * Author: Dung Nguyen Tien (Chris)
*/

public static void InsertionSort(List<int> arr)
{
	for (var i = 1; i < arr.Count; i++)
	{
		var index = BinarySearch(new List<int>(arr.GetRange(0, i - 1)), arr[i]);

		if (index != i)
		{
			var tmp = arr[i];
			arr.RemoveAt(i);
			arr.Insert(index, tmp);
		}
	}
}

public static int BinarySearch(List<int> arr, int target)
{
	if (arr.Count == 0) return 0;

	var left = 0;
	var right = arr.Count - 1;

	while (left < right)
	{
		var mid = (left + right) / 2;

		if (arr[mid] < target) left = mid + 1;
		else right = mid;
	}

	if (arr[left] < target) return left + 1;
	return left;
}

File: Common Pattern Solution/Sort/InsertionSort.js
/*
 * Problem: Insertion sort
 * Author: Dung Nguyen Tien (Chris)
*/
function insertionSort(arr) {
  for (var i = 1; i < arr.length; i++) {
    var index = binarySearch(arr.slice(0, i), arr[i]);
    if (index != i) {
      var tmp = arr.splice(i, 1)[0];
      arr.splice(index, 0, tmp);
    }
  }
}

function binarySearch(arr, target) {
  var left = 0;
  var right = arr.length - 1;
  
  while(left < right) {
    var mid = Math.floor((left + right) / 2);
    if (arr[mid] < target) left = mid + 1;
    else right = mid;
  }
  
  if (arr[left] < target) return left + 1;
  return left;
}

var arr = [1,5,3,7,8,3, 0];
insertionSort(arr);
document.writeln(JSON.stringify(arr));